### PR TITLE
refactor: intergrate new tesseract schema and update json rpc calls 

### LINKED ISSUE #42: Intergrate tesseract changes
## Description
Implement the modifications associated with MOI Protocol version 0.6.0.

**Note**: This document is a preliminary issue.

## Changes

- [x] #44
  - `Interaction`, `Receipt` and `Tesseract` schema changed. [Click here](https://www.notion.so/sarvalabs/Tesseract-6716f6f4282b439f8e222530f7104104) to know about the changes.
  - The fields `state_hashes` and `context_hashes` have been consolidated into a singular field named `hashes`.
- [x] `moi.InteractionByTesseract`
  - The address field is no longer `obligatory`. Utilization of `tesseract_hash` and `ix_index` for interaction retrieval.
- [x] `moi.GetTesseract`
  - Modification enables fetching Tesseract using tesseract_hash without the necessity of providing an address.

### PR DESCRIPTION
**Pull request by sarvalabs-harshrastogi**
### Description

This pull request updates the code structure to integrate new Tesseract structure changes and update the JSON-RPC calls.

- closes #42 #44 

### Changes include
- [ ] Bugfix (non-breaking change that solves an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

### Checklist
- [X] I have assigned this PR to myself
- [X] I have added at least 1 reviewer
- [X] I have tested this code
- [ ] I have updated the README and other relevant documents (guides...)

### CODE REVIEW

File: packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts [hunk 1]
```diff
@@ -4,18 +4,21 @@ import { JsonRpcProvider } from "../src/jsonrpc-provider";
...
-  const address = "0x898ca25ac7a51a36894b9c9f55ec6212500dd8e0c01f6591f0eb9f5b0bc84655";
-  const mnemonic = "hockey airport rather chef nasty shrimp tragic embrace olive another own hen";
-    let provider: JsonRpcProvider;
+    const address = ADDRESS;
+    const provider = new JsonRpcProvider(HOST);
     let ixHash: string;
     let signer: Signer;
     let ixReceipt: InteractionReceipt;
     let nextNonce = 0;
+    const supply = Math.floor(Math.random() * 1000);
```
Comment (sarvalabs-gokul) [inline packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts:18]: nitpick, move the supply to the top, and place it with other constants.

File: packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts [hunk 2]
```diff
@@ -24,10 +27,10 @@ describe("Test JsonRpcProvider Query Calls", () => {
         fuel_limit: 200,
         payload: {
             standard: AssetStandard.MAS0,
-            symbol: "TESTING",
-            supply: 1248577
+            symbol: "TEST #" + Math.floor(Math.random() * 1000),
```
Comment (sarvalabs-gokul) [inline packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts:30]: nitpick, can we have utility function to generate random symbol?

File: packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts [hunk 3]
```diff
@@ -123,6 +126,23 @@ describe("Test JsonRpcProvider Query Calls", () => {
       });
     });
 
+    describe("getInteractionByTesseract", () => {
+      it('should return the interaction by tesseract', async () => {
+        const interactions = await provider.getInteractionByTesseract(address, undefined, "0x0");
```
Comment (sarvalabs-gokul) [inline packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts:131]: height should be of type number.

File: packages/js-moi-providers/dist/abstract-provider.d.ts [hunk 4]
```diff
@@ -10,9 +10,11 @@ export declare abstract class AbstractProvider {
     abstract getBalance(address: string, assetId: string, options?: Options): Promise<number | bigint>;
     abstract getContextInfo(address: string, options?: Options): Promise<ContextInfo>;
     abstract getTesseract(address: string, with_interactions: boolean, options?: Options): Promise<Tesseract>;
+    abstract getTesseract(with_interactions: boolean, options?: Options): Promise<Tesseract>;
     abstract getTDU(address: string, options?: Options): Promise<TDU[]>;
     abstract getInteractionByHash(ixHash: string): Promise<Interaction>;
     abstract getInteractionByTesseract(address: string, options?: Options, ix_index?: string): Promise<Interaction>;
+    abstract getInteractionByTesseract(options: Options, ix_index?: string): Promise<Interaction>;
```
Comment (sarvalabs-gokul) [inline packages/js-moi-providers/dist/abstract-provider.d.ts:17]: ix_index should be of type `number`.

File: packages/js-moi-utils/dist/index.d.ts [hunk 5]
```diff
@@ -1,17 +1,16 @@
...
-export * from "./object";
-export * from "./hex";
-export * from "./json";
-export { Hex } from "../types/hex";
-export { AssetCreationReceipt, AssetMintOrBurnReceipt, LogicDeployReceipt, LogicInvokeReceipt } from "../types/receipt";
-export { Tesseract } from "../types/tesseract";
-export { Interaction } from "../types/interaction";
+export { Interaction, Participants, Participant } from "../types/interaction";
 export { Receipt } from "../types/receipt";
+export { Tesseract } from "../types/tesseract";
```
Comment (sarvalabs-gokul) [inline packages/js-moi-utils/dist/index.d.ts:16]: nitpick, order the export statements neatly

**Review by sarvalabs-gokul — CHANGES_REQUESTED** [review]

File: docs/source/utilities.rst [hunk 6]
```diff
@@ -16,6 +16,32 @@ to ensure accuracy, security, and convenience throughout the development process
...
+
+**Participant**
+
+The ``Participant`` interface represents a participant object. It has the following properties:
+
+* ``address`` - ``string``: The address of the participant.
+* ``height`` - ``string``: The height of the participant.
+* ``transitive_link`` - ``string``: The transitive link for the participant.
+* ``prev_context`` - ``string``: The previous context for the participant.
+* ``latest_context`` - ``string``: The latest context for the participant.
```
Comment (sarvalabs-harshrastogi) [inline docs/source/utilities.rst:37]: update description

File: packages/js-moi-providers/dist/base-provider.d.ts [hunk 7]
```diff
@@ -72,6 +72,15 @@ export declare class BaseProvider extends AbstractProvider {
      * @throws Error if there is an error executing the RPC call.
      */
     getInteractionByTesseract(address: string, options?: Options, ix_index?: string): Promise<Interaction>;
+    /**
+     * Retrieves the interaction information for the specified tesseract options.
+     *
+     * @param options - The tesseract options. (optional)
+     * @param ix_index - The index of the interaction to retrieve.
```
Comment (sarvalabs-harshrastogi) [inline packages/js-moi-providers/dist/base-provider.d.ts:79]: mark `ix_index` as optional

File: packages/js-moi-providers/dist/base-provider.d.ts [hunk 8]
```diff
@@ -194,6 +203,16 @@ export declare class BaseProvider extends AbstractProvider {
      * @throws {Error} if there is an error executing the RPC call.
      */
     getTesseract(address: string, with_interactions: boolean, options?: Options): Promise<Tesseract>;
+    /**
+      * Retrieves a Tesseract for a tesseract hash.
+      *
+      * @param {boolean} with_interactions - A boolean value indicating whether to include
+      * interactions in the Tesseract.
+      * @param {Pick<Options, 'tesseract_hash'>} options - The tesseract options. (optional)
```
Comment (sarvalabs-harshrastogi) [inline packages/js-moi-providers/dist/base-provider.d.ts:211]: update the type for `option` in js-doc

File: packages/js-moi-providers/dist/base-provider.js [hunk 9]
```diff
@@ -403,22 +417,36 @@ class BaseProvider extends abstract_provider_1.AbstractProvider {
...
-     * @param {boolean} with_interactions - A boolean value indicating whether to include
-     * interactions in the Tesseract.
-     * @param {Options} options - The tesseract options. (optional)
-     * @returns {Promise<Tesseract>} A promise that resolves to the Tesseract.
-     * @throws {Error} if there is an error executing the RPC call.
-     */
-    async getTesseract(address, with_interactions, options) {
+      * Retrieves a Tesseract for a specific address or tesseract hash.
+      * If using an address, provide the address along with whether to include interactions and optional tesseract options.
+      * If using a tesseract hash, provide a boolean indicating whether to include interactions and optional tesseract options.
```
Comment (sarvalabs-harshrastogi) [inline packages/js-moi-providers/dist/base-provider.js:422]: update js-docs

File: packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts (hunk 1 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts:21]: 4535277a344fe6e3a959a6a82168025d98f45afa

File: packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts (hunk 2 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts]: 4535277a344fe6e3a959a6a82168025d98f45afa

File: packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts [hunk 10]
```diff
@@ -169,27 +189,51 @@ describe("Test JsonRpcProvider Query Calls", () => {
     describe('getTesseract', () => {
         it('should return the latest tesseract', async () => {
           const tesseract = await provider.getTesseract(address, false);
+          const participant = tesseract.participants.find(p => p.address === address);
+
+          if (!participant) {
+            expect(participant).toBeDefined();
+            return;
+          }
```
Comment (sarvalabs-gokul) [inline packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts:200]: The if statement can be avoided

File: packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts [hunk 11]
```diff
@@ -123,6 +126,23 @@ describe("Test JsonRpcProvider Query Calls", () => {
...
+        expect(interactions).toBeDefined();
+        expect(interactions.hash).toBe(ixReceipt.ix_hash);
+      });
+
+      it("should return the interaction by tesseract without address", async () => {
+        const tesseract = await provider.getTesseract(address, true);
+        const interaction = await provider.getInteractionByTesseract({ tesseract_hash: tesseract.hash }, "0x0");
+
+        expect(interaction).toBeDefined();
+        expect(interaction.ts_hash).toBe(tesseract.hash);
```
Comment (sarvalabs-gokul) [inline packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts:145]: replace this with not null, and repeat the same for all the other occurrences.

**Review by sarvalabs-gokul — CHANGES_REQUESTED** [review]

File: packages/js-moi-providers/dist/abstract-provider.d.ts (hunk 4 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-providers/dist/abstract-provider.d.ts]: daf131566af57af4d2f8620381542beb36985b16

File: packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts (hunk 3 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts]: daf131566af57af4d2f8620381542beb36985b16

File: packages/js-moi-utils/dist/index.d.ts (hunk 5 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-utils/dist/index.d.ts]: 68dbb0137444a2ca86b44860f6459f4dba7c4ef6

File: docs/source/utilities.rst [hunk 12]
```diff
@@ -83,88 +110,37 @@ The ``Receipt`` interface represents an interaction receipt. It has the followin
...
 * ``hash`` - ``string``: The hash of the tesseract.
+* ``interactions_hash`` - ``string``: The hash of the interactions.
+* ``receipts_hash`` - ``string``: The hash of the receipts.
+* ``epoch`` - ``string``: The hex value represents the fixed time slot.
+* ``time_stamp`` - ``string``: The hex value represents the ICS request time.
+* ``operator`` - ``string``: The Krama ID of the operator.
+* ``fuel_used`` - ``string``: The hex value representing fuel used in the tesseract formation.
+* ``fuel_limit`` - ``string``: The hex value representing fuel limit of the tesseract.
+* ``consensus_info`` - ``ConsensusInfo``: The object representing the consensus information.
+* ``seal`` - ``string``: The hex value representing the The signature of node which executed the tesseract.
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply docs/source/utilities.rst:141]: f615541086d251617c1ff1b488118b0c8c15d37d

File: docs/source/utilities.rst (hunk 6 above)
Comment (sarvalabs-harshrastogi) [reply docs/source/utilities.rst]: f615541086d251617c1ff1b488118b0c8c15d37d

File: packages/js-moi-providers/dist/base-provider.d.ts (hunk 8 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-providers/dist/base-provider.d.ts]: f615541086d251617c1ff1b488118b0c8c15d37d

File: packages/js-moi-providers/dist/base-provider.js (hunk 9 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-providers/dist/base-provider.js]: f615541086d251617c1ff1b488118b0c8c15d37d

File: packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts (hunk 10 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts:200]: ff1526080cac6383d6521a80b129d8a610dc7f99

File: packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts (hunk 11 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-providers/__tests__/jsonrpc-provider.test.ts]: ff1526080cac6383d6521a80b129d8a610dc7f99

**Review by sarvalabs-gokul — APPROVED** [review]

**Pull request merged**